src/ring_theory/graded_algebra/homogeneous_ideal.lean

```diff
@@ -0,0 +1,296 @@
+subset_Inter_iff.symm
+
+lemma is_homogeneous_ideal.exists_iff_eq_span :
+  (∃ (S : set (homogeneous_submonoid A)), I = ideal.span (coe '' S)) ↔
+  I = ideal.span {x | x ∈ I ∧ is_homogeneous A x} :=
+-- get rid of the messy subtypes and set coercions
+suffices (∃ s : set R, s ⊆ set_of (is_homogeneous A) ∧ I = ideal.span s) ↔
+  I = ideal.span (I ∩ set_of (is_homogeneous A)),
+from (subtype.exists_set_subtype _).trans this,
+begin
+  split,
+  { rintros ⟨s, hs, rfl⟩,
+    apply le_antisymm,
+    { exact ideal.span_mono (subset_inter (ideal.subset_span) hs) },
+    { exact ideal.span_le.2 (inter_subset_left _ _) } },
+  { intros hI,
+    exact ⟨(I : set R) ∩ set_of (is_homogeneous A), inter_subset_right _ _, hI⟩, }
+end
+
+lemma mul_homogeneous_element_mem_of_mem
+  {I : ideal R} (r x : R) (hx₁ : is_homogeneous A x) (hx₂ : x ∈ I) (j : ι) :
+  graded_algebra.proj A j (r * x) ∈ I :=
+begin
+  letI : Π (i : ι) (x : A i), decidable (x ≠ 0) := λ _ _, classical.dec _,
+  rw [←graded_algebra.sum_support_decompose A r, finset.sum_mul, linear_map.map_sum],
+  apply ideal.sum_mem,
+  intros k hk,
+  obtain ⟨i, hi⟩ := hx₁,
+  have mem₁ : (graded_algebra.decompose A r k : R) * x ∈ A (k + i) := graded_monoid.mul_mem
+    (submodule.coe_mem _) hi,
+  rw [graded_algebra.proj_apply, graded_algebra.decompose_of_mem A mem₁,
+    coe_of_submodule_apply A, submodule.coe_mk],
+  split_ifs,
+  { exact I.mul_mem_left _ hx₂ },
+  { exact I.zero_mem },
+end
+
+lemma is_homogeneous_ideal_iff_eq :
+  is_homogeneous_ideal A I ↔ I = ideal.span {x | x ∈ I ∧ is_homogeneous A x} :=
+⟨ λ hI, begin
+  letI : Π (i : ι) (x : A i), decidable (x ≠ 0) := λ _ _, classical.dec _,
+  ext, split; intro hx,
+  { rw ←graded_algebra.sum_support_decompose A x,
+    refine ideal.sum_mem _ _,
+    intros j hj,
+    rw ideal.mem_span, intros J HJ,
+    refine HJ ⟨hI j hx, j, submodule.coe_mem _⟩, },
+  { rw [ideal.mem_span] at hx,
+    apply hx,
+    exact inter_subset_left _ _, },
+  end,
+  λ hI, begin
+    intros i r hr,
+    rw ←graded_algebra.proj_apply,
+    rw [ideal.span, finsupp.span_eq_range_total] at hI,
+    rw hI at hr,
+    obtain ⟨s, rfl⟩ := hr,
+    simp_rw [finsupp.total_apply, finsupp.sum, linear_map.map_sum, smul_eq_mul],
+    refine ideal.sum_mem I _,
+    rintros ⟨j, ⟨hj₁, hj₂⟩⟩ hj₃,
+    exact mul_homogeneous_element_mem_of_mem _ _ _ hj₂ hj₁ _,
+  end ⟩
+
+lemma is_homogeneous_ideal_iff_exists :
+  is_homogeneous_ideal A I ↔ ∃ (S : set (homogeneous_submonoid A)), I = ideal.span (coe '' S) :=
+by rw [is_homogeneous_ideal.exists_iff_eq_span, is_homogeneous_ideal_iff_eq]
+
+end is_homogeneous_ideal_defs
+
+section operations
+
+open set_like direct_sum set
+open_locale big_operators pointwise
+
+variables {ι R : Type*} [comm_ring R] [decidable_eq ι] [add_comm_monoid ι]
+variables (A : ι → submodule R R) [graded_algebra A]
+variable (I : ideal R)
+
+/--We collect all homogeneous ideal into a type.-/
```

Comment:
This comment only makes sense in this file. Please write a standalone docstring that explains this type in a way that is useful outside the context of this file.